graphix/pattern.py

```diff
@@ -201,6 +209,114 @@ def __eq__(self, other: Pattern) -> bool:
+                stacklevel=2,
+            )
+            raise Exception("`pdflatex` call did not succeed: see `latex_error.log`.") from exc
+
+        base = Path(tmpdirname) / tmpfilename
+        try:
+            subprocess.run(
+                ["pdftocairo", "-singlefile", "-png", "-q", base.with_suffix(".pdf"), base],
+                check=True,
+            )
+        except (OSError, subprocess.CalledProcessError) as exc:
+            message = "`pdftocairo` failed to produce an image."
+            warnings.warn(message, stacklevel=2)
+            raise Exception(message) from exc
+
+        def _trim(image) -> PIL.Image.Image:
+            """Trim a PIL image and remove white space."""
+            background = PIL.Image.new(image.mode, image.size, image.getpixel((0, 0)))
+            diff = PIL.ImageChops.difference(image, background)
+            diff = PIL.ImageChops.add(diff, diff, 2.0, -100)
+            bbox = diff.getbbox()
+            if bbox:
+                image = image.crop(bbox)
+            return image
+
+        return _trim(PIL.Image.open(base.with_suffix(".png")))
+
+    def to_latex(self) -> str:
+        """Return a string containing the latex representation of the pattern."""
+        output = io.StringIO()
+        for cmd in self.__seq:
+            output.write(command_to_latex(cmd))
+            output.write("\n")
+
+        contents = output.getvalue()
+        output.close()
+        return contents
+
+    def _to_latex_document(self) -> str:
+        """Generate a latex document with the latex representation of the pattern written in it."""
+        header_1 = r"\documentclass[border=2px]{standalone}" + "\n"
+
+        header_2 = r"""
+\usepackage{graphicx}
+
+\begin{document}
+"""
+
+        output = io.StringIO()
+        output.write(header_1)
+        output.write(header_2)
+
+        output.write(self.to_latex())
+
+        output.write("\n\\end{document}")
+        contents = output.getvalue()
+        output.close()
+
+        return contents
+
+    def to_png(self) -> PIL.Image.Image:
+        """Generate a PNG image of the latex representation of the pattern."""
+        tmpfilename = "pattern"
+
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            tmppath = Path(tmpdirname) / tmpfilename
+            tmppath = tmppath.with_suffix(".tex")
+
+            with open(tmppath, "w") as latex_file:
+                contents = self._to_latex_document()
+                latex_file.write(contents)
+
+            return self._latex_file_to_image(tmpdirname, tmpfilename)
+
+    def __str__(self) -> str:
+        """Return a string representation of the pattern."""
+        return "\n".join([command_to_str(cmd) for cmd in self.__seq])
+
+    def to_unicode(self) -> str:
+        """Return the unicode string representation of the pattern."""
+        return "".join([command_to_unicode(cmd) for cmd in self.__seq])
```

Comment:
I think inserting a space between commands (`" "` instead of `""`) would be the output more readable.